src/libraries/System.Runtime.Serialization.BinaryFormat/src/System/Runtime/Serialization/BinaryFormat/SerializationRecord.cs

```diff
@@ -0,0 +1,58 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+
+using System.Diagnostics;
+
+namespace System.Runtime.Serialization.BinaryFormat;
+
+/// <summary>
+/// Abstract class that represents the serialization record.
+/// </summary>
+/// <remarks>
+///  <para>
+///   Every instance returned to the end user can be either <seealso cref="PrimitiveTypeRecord{T}"/>,
+///   a <seealso cref="ClassRecord"/> or an <seealso cref="ArrayRecord"/>.
+///  </para>
+/// </remarks>
+[DebuggerDisplay("{RecordType}, {ObjectId}")]
+#if SYSTEM_RUNTIME_SERIALIZATION_BINARYFORMAT
+public
+#else
+internal
+#endif
+abstract class SerializationRecord
+{
+    internal const int NoId = 0;
+
+    internal SerializationRecord() // others can't derive from this type
+    {
+    }
+
+    public abstract RecordType RecordType { get; }
+
+    public virtual int ObjectId => NoId;
```

Comment:
> I assume you're avoiding `int?` because none of the record types which lack identifiers are public

This is correct. Moreover, it's impossible to get them using the public APIs, as:
- `ObjectNullMultiple256Record`, `ObjectNullMultipleRecord` and `ObjectNullRecord` are simply mapped to `null`(s) in all public APIs
- `MessageEndRecord` is never returned to the user or added to record map

Reply:
> If it's only one or two, I'd leave this as abstract and let it be clear in those other record types that they have no ID.

I am going to give it a try